### PR TITLE
Adds etcd storage

### PR DESCRIPTION
**Pull request by JosephSalisbury**

### CODE REVIEW

**Comment by JosephSalisbury** [comment]
```
13:57:31 ~/go/src/github.com/giantswarm/etcdstorage (initial)
$ go test -v $(glide novendor) -tags=integration
=== RUN   Test_Storage
--- PASS: Test_Storage (0.36s)
PASS
ok  	github.com/giantswarm/etcdstorage	0.378s
```

with local etcd instance

File: error.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+var invalidConfigError = microerror.New("invalid config")
+
+// IsInvalidConfig asserts invalidConfigError.
+func IsInvalidConfig(err error) bool {
+	return microerror.Cause(err) == invalidConfigError
+}
+
+var multipleValuesError = microerror.New("multiple values")
```
Comment (JosephSalisbury) [inline error.go:14]: There is an argument that this should be in `microstorage`, but I don't see the need right now.

**Comment by JosephSalisbury** [comment]
I'm definitely putting a lot of face in `storagetest`, but that's kind of the point, right? :D

**Comment by JosephSalisbury** [comment]
I'll delete the etcd storage in microkit after this.

File: etcd.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+func (s *Service) Put(ctx context.Context, key, value string) error {
+	ctx, cancel := context.WithTimeout(ctx, s.timeout)
+	defer cancel()
+
+	var err error
+
+	key, err = microstorage.SanitizeKey(key)
+	if err != nil {
+		return microerror.Mask(err)
+	}
```
Comment (kopiczko) [inline etcd.go:96]: It makes sense to perform sanitisation before setting timeout on context.

File: etcd_test.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	config := DefaultConfig()
+	config.EtcdClient = etcdClient
+
+	storage, err := New(config)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	storagetest.Test(t, storage)
+}
```
Comment (kopiczko) [inline etcd_test.go:30]: Looks good. Did you see some failing which you had to fix?

Comment (JosephSalisbury) [reply etcd_test.go:30]: I had issues with key sanitisation and returning internal errors, but nothing major.

File: etcd.go (hunk 2 above)
Comment (JosephSalisbury) [reply etcd.go:96]: Yeah, fine. Will fix.

**Comment by kopiczko** [comment]
> I'll delete the etcd storage in microkit after this.

Be careful. I think you should hold your horses. I'd see how migration to microstorage goes first.

File: etcd.go (hunk 2 above)
Comment (JosephSalisbury) [reply etcd.go:96]: Fixed.

File: etcd.go [hunk 4]
```diff
@@ -69,16 +69,16 @@ type Service struct {
 }
 
 func (s *Service) Create(ctx context.Context, key, value string) error {
-	ctx, cancel := context.WithTimeout(ctx, s.timeout)
-	defer cancel()
-
```
Comment (kopiczko) [inline etcd.go]: diff is wrong, you moved sanitisation up, not context wrapping down :P

**Review by kopiczko — APPROVED** [review]
Thanks! <3

**Pull request merged**